test/integration/test_server.cpp

```diff
@@ -594,9 +612,122 @@ TEST_CASE("CreateTable with multiple primary keys", "[integration]") {
       ::fivetran_sdk::DescribeTableResponse response;
       auto status = service.DescribeTable(nullptr, &request, &response);
       REQUIRE_NO_FAIL(status);
-      REQUIRE(response.table().columns().size() == 2);
+      REQUIRE(response.table().columns().size() == 5);
+
+      REQUIRE(response.table().columns(0).name() == "id1");
+      REQUIRE(response.table().columns(1).name() == "id2");
+      REQUIRE(response.table().columns(2).name() == "text");
+      REQUIRE(response.table().columns(3).name() == "_fivetran_deleted");
+      REQUIRE(response.table().columns(4).name() == "_fivetran_synced");
     }
   }
+
+  // test connection needs to be created after table creation to avoid stale
+  // catalog
+  auto con = get_test_connection(token);
+  {
+    // insert rows
+    ::fivetran_sdk::WriteBatchRequest request;
+    (*request.mutable_configuration())["motherduck_token"] = token;
+    (*request.mutable_configuration())["motherduck_database"] = "fivetran_test";
+    define_test_multikey_table(request, table_name);
+    const std::string filename = "multikey_table_upsert.csv";
+    const std::string filepath = TEST_RESOURCES_DIR + filename;
+
+    request.add_replace_files(filepath);
+
+    ::fivetran_sdk::WriteBatchResponse response;
+    auto status = service.WriteBatch(nullptr, &request, &response);
+    REQUIRE_NO_FAIL(status);
+  }
+
+  {
+    // check inserted rows
+    auto res = con->Query("SELECT id1, id2, text FROM " + table_name +
+                          " ORDER BY id1, id2");
+    REQUIRE_NO_FAIL(res);
+    REQUIRE(res->RowCount() == 3);
+    REQUIRE(res->GetValue(0, 0) == 1);
+    REQUIRE(res->GetValue(1, 0) == 100);
+    REQUIRE(res->GetValue(2, 0) == "first row");
+
+    REQUIRE(res->GetValue(0, 1) == 2);
+    REQUIRE(res->GetValue(1, 1) == 200);
+    REQUIRE(res->GetValue(2, 1) == "second row");
+
+    REQUIRE(res->GetValue(0, 2) == 3);
+    REQUIRE(res->GetValue(1, 2) == 300);
+    REQUIRE(res->GetValue(2, 2) == "third row");
+  }
+
+  {
+    // update
+    ::fivetran_sdk::WriteBatchRequest request;
```

Comment:
Out of curiosity where is this fivetran code coming from? Are we generating it from https://github.com/fivetran/fivetran_sdk ?

Reply:
Nvm just read through the build files and I see!